### PR TITLE
Add padding as an option to torch.cat

### PR DESCRIPTION
**Pull request by yya007-zz**
This PR is let torch.cat support padding.

Example:
>>> x = torch.ones(2, 3)
>>> y = torch.zeros(1, 4)
>>> torch.cat([x, y], dim=0, pad=True, pad_value=0.5)

tensor([[1., 1., 1., 0.5],
              [1., 1., 1., 0.5],
              [0., 0., 0., 0.]])

See T33342447

### CODE REVIEW

**Review by apaszke — CHANGES_REQUESTED** [review]
Extra fixes needed for the JIT

File: aten/src/TH/generic/THTensorMoreMath.cpp [hunk 1]
```diff
@@ -1290,35 +1290,39 @@ void THTensor_(triu)(THTensor *r_, THTensor *t, int64_t k)
...
   THTensor* inputs[2];
   inputs[0] = ta;
   inputs[1] = tb;
-  THTensor_(catArray)(r_, inputs, 2, dimension);
+  THTensor_(catArray)(r_, inputs, 2, dimension, pad, pad_value);
 }
 
-void THTensor_(check_shape_except_dim)(THTensor *first, THTensor *second, int dimension);
-inline void THTensor_(check_shape_except_dim)(THTensor *first, THTensor *second, int dimension)
+void THTensor_(check_shape_except_dim)(THTensor *first, THTensor *second, int dimension, scalar_t pad);
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorMoreMath.cpp]: why is pad a scalar_t?  You are checking whether padding is on, not what the value of the padding is.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -736,11 +736,11 @@ def parse_kwargs(desc):
 
 add_docstr(torch.cat,
            r"""
-cat(seq, dim=0, out=None) -> Tensor
+cat(seq, dim=0, pad=False, pad_value=0, out=None) -> Tensor
 
 Concatenates the given sequence of :attr:`seq` tensors in the given dimension.
 All tensors must either have the same shape (except in the concatenating
-dimension) or be empty.
+dimension) or be empty. Otherwise, :attr:`pad` must be set true to force cat.
```
Comment (gchanan) [inline torch/_torch_docs.py]: A couple things here:
1) We should remove the part about "or be empty".  This is legacy behavior that people shouldn't rely on (this is not related to your PR, but since you are touching it, it would be good to fix up).
2) It's not clear to me what "to force cat" means or how the padding is done.  I would just split this paragraph in two.  The first says: "If padding is set to false,...", the second "If padding is set to true..." and explain those cases separately.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -752,6 +752,9 @@ def parse_kwargs(desc):
         Non-empty tensors provided must have the same shape, except in the
         cat dimension.
     dim (int, optional): the dimension over which the tensors are concatenated
+    pad (bool, optional): add paddings to expand input tensors to same shape
```
Comment (gchanan) [inline torch/_torch_docs.py]: this needs more explanation of how the padding is done; you can refer to the paragraph split I mention above.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -772,6 +775,18 @@ def parse_kwargs(desc):
              -1.0969, -0.4614],
             [-0.1034, -0.5790,  0.1497, -0.1034, -0.5790,  0.1497, -0.1034,
              -0.5790,  0.1497]])
+    >>> x = torch.ones(2, 3, device='cpu')
```
Comment (gchanan) [inline torch/_torch_docs.py]: put a comment here saying with padding=True.

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -772,6 +775,18 @@ def parse_kwargs(desc):
...
+    >>> torch.cat([x, y], dim=0, pad=True, pad_value=0.5)
+    tensor([[1.0000, 1.0000, 1.0000, 0.5000],
+            [1.0000, 1.0000, 1.0000, 0.5000],
+            [0.0000, 0.0000, 0.0000, 0.0000]])
+    >>> x = torch.ones(2, 3, device='cuda')
+    >>> y = torch.zeros(1, 4, device='cuda')
+    >>> torch.cat([x, y], dim=0, pad=True, pad_value=0.5)
+    tensor([[1.0000, 1.0000, 1.0000, 0.5000],
+            [1.0000, 1.0000, 1.0000, 0.5000],
+            [0.0000, 0.0000, 0.0000, 0.0000]])
```
Comment (gchanan) [inline torch/_torch_docs.py]: wait -- that doesn't create a cuda tensor?  (You don't need this example since it's redundant, but it's weird that it's not a cuda tensor).

File: aten/src/TH/generic/THTensorMoreMath.cpp [hunk 6]
```diff
@@ -1402,7 +1424,17 @@ void THTensor_(catArray)(THTensor *result, THTensor **inputs, int numInputs, int
...
-        THTensor_(narrow)(nt, NULL, dimension, offset, dimSize);
+        if (pad) {
+          for (int dim = 0; dim < nDims; dim++) {
+            if (dimension == dim) {
+              THTensor_(narrow)(nt, NULL, dimension, offset, dimSize);
+            }else{
+              THTensor_(narrow)(nt, NULL, dim, 0, inputs[j]->size(dim));
+            }
+          }
+        }else{
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorMoreMath.cpp]: nit: spacing.

**Review by gchanan — COMMENTED** [review]
can you add autograd tests?
